workflows/prognostic_c48_run/runtime/emulator.py

```diff
@@ -45,102 +71,120 @@ class Config:
-    emulator_prefix: str = "emulator_"
+    label: str
     diagnostic_variables: Set[str] = dataclasses.field(default_factory=set)
     timestep: float = 900
 
-    def __post_init__(self: "PrognosticAdapter"):
-        self.emulator = get_xarray_emulator(self.config.emulator)
-
     @property
     def monitor(self) -> Monitor:
         return Monitor.from_variables(
             self.diagnostic_variables, self.state, self.timestep
         )
 
     @property
-    def inputs_to_save(self) -> Set[str]:
-        return set(strip_prefix(self.emulator_prefix, self.diagnostic_variables))
+    def prefix(self) -> str:
+        return self.label + "_"
 
-    def emulate(self, name: str, func: Step) -> Diagnostics:
-        inputs = {key: self.state[key] for key in self.emulator.input_variables}
+    @property
+    def inputs_to_save(self) -> Set[str]:
+        return set(strip_prefix(self.prefix, self.diagnostic_variables))
 
+    def transform(self, name: str, func: Step) -> Diagnostics:
+        inputs: State = {key: self.state[key] for key in self.model.input_variables}
         inputs_to_save: Diagnostics = {
-            self.emulator_prefix + key: self.state[key] for key in self.inputs_to_save
+            self.prefix + key: self.state[key] for key in self.inputs_to_save
         }
+
         before = self.monitor.checkpoint()
         diags = func()
-        change_in_func = self.monitor.compute_change(name, before, self.state)
-
-        if self.config.train:
-            self.emulator.partial_fit(inputs, self.state)
-        emulator_prediction = self.emulator.predict(inputs)
+        change_due_to_func = self.monitor.compute_change(name, before, self.state)
 
-        emulator_after: MutableMapping[Hashable, xr.DataArray] = {
-            DELP: before[DELP],
-            **emulator_prediction,
-        }
+        self.model.partial_fit(inputs, self.state)
 
-        # insert state variables not predicted by the emulator
+        prediction = self.model.predict(inputs)
         for v in before:
-            if v not in emulator_after:
-                emulator_after[v] = self.state[v]
-
-        changes = self.monitor.compute_change("emulator", before, emulator_after)
+            if v not in prediction:
+                prediction[v] = self.state[v]
 
-        if self.config.online:
-            updated_state = where_masked(
-                self.state,
-                emulator_prediction,
-                compute_mask=get_mask(
-                    self.config.mask_kind, self.config.ignore_humidity_below
-                ),
-            )
-            self.state.update(updated_state)
-
-        return {**diags, **inputs_to_save, **changes, **change_in_func}
+        change_due_to_ml = self.monitor.compute_change(self.label, before, prediction)
+        self.model.apply(prediction, self.state)
+        change_due_to_applied_ml = self.monitor.compute_change(
+            f"applied_{self.label}", before, self.state
+        )
```

Comment:
We cannot promise that these are actually the tendencies applied (since we could be inside of another `StepTransfomer`).  I would use another `monitor` to do this on the higher level of the code. Here is an example of how to do this:
```
--- a/workflows/prognostic_c48_run/runtime/loop.py
+++ b/workflows/prognostic_c48_run/runtime/loop.py
@@ -224,7 +224,7 @@ class TimeLoop(
         elif self._emulate is not None and self._prescribe_tendency is None:
             return self._emulate(name, func)
         else:
-            return self.monitor(name, func)
+            return func
 
     def _get_prephysics_stepper(
         self, config: UserConfig, hydrostatic: bool
@@ -471,7 +471,12 @@ class TimeLoop(
                 self._step_prephysics,
                 self._compute_physics,
                 self._apply_postphysics_to_physics_state,
-                self.emulate_or_prescribe_tendency("fv3_physics", self._apply_physics),
+                self.monitor(
+                    "applied_physics",
+                    self.emulate_or_prescribe_tendency(
+                        "fv3_physics", self._apply_physics
+                    ),
+                ),
                 self._compute_postphysics,
                 self.monitor("python", self._apply_postphysics_to_dycore_state),
             ]:
```

Reply:
Nice, yah that looks good.

Reply:
Likewise, the `fv3_physics` monitoring could be removed from this code as well, and implemented like this
```
self.monitor("applied_physics", self.emulate("emulator", self.monitor("fv3_physics", self._apply_physics)))
```

Reply:
Although i would prefer to not apply
```
-            return self.monitor(name, func)
+            return func
```
in order to maintain backwards compatibility with naming. e.g. in a non-emulation run still want to be able to output `...due_to_fv3_physics`

Reply:
It would just be equal to `applied_physics` in this case.